src/main/java/apoc/cypher/Timeboxed.java

```diff
@@ -24,26 +24,85 @@
  */
 public class Timeboxed {
 
-
     @Context
     public GraphDatabaseService db;
-    @Context
-    public KernelTransaction tx;
 
     @Context
     public Log log;
 
+    private static Map<String,Object> POISON = Collections.singletonMap("__magic", "POISON");
+
     @Procedure
     @Description("apoc.cypher.runTimeboxed('cypherStatement',{params}, timeout) - abort statement after timeout ms if not finished")
     public Stream<MapResult> runTimeboxed(@Name("cypher") String cypher, @Name("params") Map<String, Object> params, @Name("timeout") long timeout) {
 
+        final BlockingQueue<Map<String, Object>> queue = new ArrayBlockingQueue<>(100);
+        final AtomicReference<Transaction> txAtomic = new AtomicReference<>();
+
+        // run query to be timeboxed in a separate thread to enable proper tx termination
+        // if we'd run this in current thread, a tx.terminate would kill the transaction the procedure call uses itself.
+        Pools.DEFAULT.submit(() -> {
+            try (Transaction tx = db.beginTx()) {
+                txAtomic.set(tx);
+                Result result = db.execute(cypher, params == null ? Collections.EMPTY_MAP : params);
+                while (result.hasNext()) {
+                    final Map<String, Object> map = result.next();
+                    queue.put(map);
+                }
+                queue.put(POISON);
+                tx.success();
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            } catch (TransactionTerminatedException e) {
+                log.warn("query " + cypher + " has been terminated");
+            } finally {
+                txAtomic.set(null);
+            }
+        });
+
+        //
         Pools.SCHEDULED.schedule(() -> {
-            String txString = tx == null ? "<null>" : tx.toString();
-            log.warn("marking " + txString + " for termination");
-            tx.markForTermination(Status.Transaction.Terminated);
+            Transaction tx = txAtomic.get();
+            if (tx==null) {
+                log.error("oops, tx is null, maybe other thread already finished?");
```

Comment:
is that an error?

Reply:
this could potentially happen if you set the timeout value so small that the other transaction has not yet been started. I consider it an error. Will amend the message to something more meaningful